src/maturin/springs.clj

```diff
@@ -0,0 +1,62 @@
+;; create xs
+(defn gen-xs [n]
+    (take n (repeatedly #(rand 10))))
+  
+;; create linear regression mock data
+(defn gen-ys [xs slope intercept]
+    (map #(+ (d/sample (d/normal 0 1)) intercept (* % slope)) xs))
+
+;; return linear function
+(defn linear-func [slope intercept]
+    (fn [x] (+ intercept (* slope x)))) 
+
+;; lagrangian for the springs-rod system, regression case:
+;; springs constrained to have the same x-coordiate on both axes = minimize MSE
+(defn L-regression
+    [xs ys k]
+    (fn [[_ [slope intercept] _]]
+        (let [preds (map (linear-func slope intercept) xs)
```

Comment:
these are definitely generalized coordinates :)

If you wanted to decompose this further, you could make a "spring" lagrangian, that takes, for position, a sequence of spring-stretches only... and then compose that with a transformation that does what you do internally, maps the

```clojure
(defn line->offsets [xs ys]
  (fn [[slope intercept]]
    (let [preds (map (linear-func slope intercept) xs)]
      (mapv - preds ys))))
```

Then

```clojure
(defn L-regression [xs ys k]
  (compose (L-springs k) (F->C (line->offsets xs ys))
```

should work! Just an idea...

Reply:
Implemented this, ran into a different problem, see comment below.